source/base/mpi.cc

```diff
@@ -197,7 +197,54 @@ namespace Utilities
                    "There is no point in communicating with ourselves."));
         }
 
+#  if DEAL_II_MPI_VERSION_GTE(2, 2)
+      // Get a binary vector from the `destinations`
+      std::vector<unsigned int> dest_vector(n_procs);
+      for (const auto &el : destinations)
+        dest_vector[el] = 1;
+
+      // Find how many processes will send to this one
+      // by reducing with sum and then scattering the
+      // results over all processes
+      unsigned int n_recv_from;
+      const int    ierr = MPI_Reduce_scatter_block(
+        &dest_vector[0], &n_recv_from, 1, MPI_UNSIGNED, MPI_SUM, mpi_comm);
 
+      AssertThrowMPI(ierr);
+
+      // Send myid to every process in `destinations` vector...
+      std::vector<MPI_Request> send_requests(destinations.size());
+      for (const auto &el : destinations)
+        MPI_Isend(&myid,
+                  1,
+                  MPI_UNSIGNED,
+                  el,
+                  1 << 26,
```

Comment:
This value is too large in general, the MPI standard requires MPI_TAG_UB (the maximum allowed tag) to be at least 32767 (2^15-1). Just pick something below that.